content/academy/puppeteer_playwright/page.md

```diff
@@ -0,0 +1,311 @@
+```
+
+Then, we can type some text into an input field with `page.type()`; passing a CSS selector as the first, and the string to input as the second parameter:
+
+```JavaScript
+// Type the query into the search box
+await page.type('input[title]', 'hello world');
+```
+
+Finally, we can press a single key by access the `keyboard` property of `page` and calling the `press()` function on it:
+
+```JavaScript
+// Press enter
+await page.keyboard.press('Enter');
+```
+
+So far, we've got this:
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+import { chromium } from 'playwright';
+
+const sleep = (secs) => new Promise((r) => setTimeout(r, secs * 1000));
+
+const browser = await chromium.launch({ headless: false });
+
+const page = await browser.newPage();
+
+await page.goto('https://google.com/');
+
+// Click the "I understand" button
+await page.click('button + button:nth-child(2)');
+
+// Type the query into the search box
+await page.type('input[title]', 'hello world');
+
+// Press enter
+await page.keyboard.press('Enter');
+
+await sleep(10);
+await browser.close();
+</marked-tab>
+<marked-tab header="Puppeteeer" lang="javascript">
+import puppeteer from 'puppeteer';
+
+const sleep = (secs) => new Promise((r) => setTimeout(r, secs * 1000));
+
+const browser = await puppeteer.launch({ headless: false });
+
+const page = await browser.newPage();
+
+await page.goto('https://google.com/');
+
+// Click the "I understand" button
+await page.click('button + button:nth-child(2)');
+
+// Type the query into the search box
+await page.type('input[title]', 'hello world');
+
+// Press enter
+await page.keyboard.press('Enter');
+
+await sleep(10);
+await browser.close();
+</marked-tab>
+```
+
+When we run it, we leave off on the results page:
+
+![Google results page reached by headless browser]({{@asset puppeteer_playwright/images/google-results.webp}})
+
+### Waiting for elements/events
+
+In a perfect world, every piece of content served on a website would be loaded instantly. We don't live in a perfect world though, and often times it can take anywhere 1/10th of a second to a few seconds to load some content onto a page. Puppeteer and Playwright don't sit around waiting for the pages to load though - if we tell it to do something with an element that hasn't rendered yet, it'll start trying to do it (which will result in nasty errors). We've got to tell it to wait.
+
+Different events can be waited for using the various `waitFor...` methods offered, while elements can be waited for with `page.waitForSelector()` in both Playwright and Puppeteer. In Playwright you can also create a locator for an element and use the `locator.waitFor()` function.
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+// Create a pointer for the first result element
+const firstResult = page.locator(':nth-match(.g a, 1)');
+
+// Wait for the first result element to be visible
+await firstResult.waitFor();
```

Comment:
I think Playwright auto-waits before clicking, so this shouldn't be necessary. Not sure about Puppeteer, but it didn't do that the last time I checked. Yet another difference that might be worth mentioning.

Reply:
You're right!